### PR TITLE
ENH: add matvec and vecmat gufuncs

### LINKED ISSUE #12348: ENH: expose matmat, matvec, vecmat ufuncs
@charris [commented](https://github.com/numpy/numpy/pull/12219#issuecomment-436412360) on PR #12219 wishing to expose the vector-matrix matrix-vector loops function from the matmul ufunc. 

We would need two ufuncs with signature `(1,n),(n,p?)->(1,p?)` and `(m?,n),(n,1)->(m?,1)`, and could reuse the `@TYPE@_matmul` loops

### PR DESCRIPTION
**Pull request by mhvk**
This PR adds new `matvec` and `vecmat` gufunc to calculate the matrix-vector and vector-matrix product, to add to plain matrix multiplication with `matmul` and the inner vector product with `vecdot`. 

Fixes #12348

Note that for complex numbers, `vecmat` is defined as `x†A`, i.e., the complex conjugate of the vector is taken. This seems to be the standard and is what we used for `vecdot` too (`x†x`). However, it is *not* what `matmul` does for vector-matrix or indeed vector-vector products. I think this is a bug in matmul, which I'm happy to fix here. But I'll post to the mailing list to get feedback.

Separately, with these functions available, in principle these could be used in `__matmul__` and the specializations in `np.matmul` removed. But that can be a separate PR (if it is wanted at all).

*Summary of [mailing list discussion](https://mail.python.org/archives/list/numpy-discussion@python.org/thread/LX36TEAENIDSSFAKS6YLXXLWVOBABSEC/)*
- Behaviour of `np.matmul` for vector-matrix should not be adjusted. Some surprise that "matrix multiplcation" deals with vectors at all.
- Less obvious consensus on what `@` should do.
- Moderate support for having special `matvec` and `vecmat` functions.

### CODE REVIEW

**Comment by mhvk** [comment]
Refactored to avoid duplication of code: `matvec` now uses the `matmul_noblas` loops internally (as does the matrix-vector implementation in `matmul` itself), and `vecmat` creates a loop that uses the loops in `vecdot`, letting their outer loop deal with the matrix dimension.

**Comment by mhvk** [comment]
@charris - since you were the one originally asking for it, what do you think of this implementation? It is now quite lightweight, since it reuses inner loops.

**Comment by mhvk** [comment]
macos failure unrelated

**Comment by seberg** [comment]
Moving milestone to 2.1, if we get this in we should change it again, but at least I don't think it needs to be prioritized.

**Comment by mhvk** [comment]
Yes, not critical for 2.0 for sure, but maybe more logical to clear the milestone rather than set it to 2.1 if there still is a chance. This PR has become fairly simple after my refactor.

**Comment by ev-br** [comment]
Out of curiousity, what is the status of this PR?

**Comment by mhvk** [comment]
I think it would still be nice to get this in!  I just rebased to make sure things still pass. @charris - as the person who originally suggested to have `matvec` and `vecmat`, what do you think?

**Comment by mhvk** [comment]
The circle-ci failure does not seem related:
```
/home/circleci/repo/doc/source/reference/arrays.promotion.rst:150: ERROR: Unexpected indentation.
/home/circleci/repo/doc/source/reference/arrays.promotion.rst:154: WARNING: Block quote ends without a blank line; unexpected unindent.
/home/circleci/repo/doc/source/reference/arrays.promotion.rst:189: ERROR: Unexpected indentation.
/home/circleci/repo/doc/source/reference/arrays.promotion.rst:191: WARNING: Block quote ends without a blank line; unexpected unindent.
/home/circleci/repo/doc/source/reference/arrays.promotion.rst:210: ERROR: Unexpected indentation.
/home/circleci/repo/doc/source/reference/arrays.promotion.rst:212: WARNING: Block quote ends without a blank line; unexpected unindent.
/home/circleci/repo/doc/source/reference/arrays.promotion.rst:31: ERROR: Too many autonumbered footnote references: only 0 corresponding footnotes available.
/home/circleci/repo/doc/source/reference/arrays.promotion.rst:31: ERROR: Unknown target name: "hist-reasons".
/home/circleci/repo/doc/source/reference/arrays.promotion.rst:68: ERROR: Unknown target name: "nep50".
source/release/notes-towncrier.rst:119: WARNING: Inline literal start-string without end-string.
/home/circleci/repo/doc/source/release/notes-towncrier.rst:119: WARNING: Inline literal start-string without end-string.
source/release/notes-towncrier.rst:119: WARNING: Inline literal start-string without end-string.
```

File: numpy/_core/code_generators/ufunc_docstrings.py
Comment (seberg) [inline numpy/_core/code_generators/ufunc_docstrings.py:3045]: Needs to be updated, and I think we should move it to the end of the intro.

File: numpy/_core/code_generators/ufunc_docstrings.py
Comment (seberg) [inline numpy/_core/code_generators/ufunc_docstrings.py:3030]: Maybe also matmul?

File: numpy/_core/code_generators/ufunc_docstrings.py
Comment (seberg) [inline numpy/_core/code_generators/ufunc_docstrings.py:2995]: It's good, but if we have an idea, I don't think "a matrix in ``x1``" is very clear about referring to stacks of matrices/vectors.

In fact it would be nice to:
* Add `matvec` and `vecmat` to `matmul` "See Also"
* Probably also to its notes on stacks (IIRC we have those).

File: numpy/_core/code_generators/ufunc_docstrings.py
Comment (seberg) [inline numpy/_core/code_generators/ufunc_docstrings.py:3038]: Maybe we can name these `Y_vectors`, name a `rotation_matrix`, and format all as with the rows below one another (like the result).
Could also call it `X_vectors = ...` and then print that repr to document it.

**Review by seberg — APPROVED** [review]
FWIW, I am still happy with the addition.  A couple of doc comments that would be nice to address.  I think the main point is actually that it would be good to expand `matmul` docs to explain that these exist and fill a gap for stacks of vectors.

File: numpy/_core/code_generators/ufunc_docstrings.py
Comment (mhvk) [reply numpy/_core/code_generators/ufunc_docstrings.py:3045]: OK, updated. I agree the location is not very logical, but kept there anyway, since `matmul` and `vecdot` have it there too.

File: numpy/_core/code_generators/ufunc_docstrings.py
Comment (mhvk) [reply numpy/_core/code_generators/ufunc_docstrings.py:3030]: Yes, makes sense. Also added `vecdot` (and the new routines to the `vecdot` and `matmul` docs).

File: numpy/_core/code_generators/ufunc_docstrings.py
Comment (mhvk) [reply numpy/_core/code_generators/ufunc_docstrings.py:2995]: OK, I tried to reword (will push up in a bit).

File: numpy/_core/code_generators/ufunc_docstrings.py
Comment (mhvk) [reply numpy/_core/code_generators/ufunc_docstrings.py:3038]: I reformatted the arrays, but kept the names (since A and v are used in the description).

**Comment by seberg** [comment]
One thing that occurred to me, it might also make sense to put these into `np.linalg`.  OTOH, since matmul and vecdot are in the main namespace, it is probably clearer there.

**Comment by mhvk** [comment]
> One thing that occurred to me, it might also make sense to put these into `np.linalg`.

I think the main namespace is good enough, for now at least. It might change if the array API adopts them.

**Comment by mhvk** [comment]
@seberg - shall we get this in? I fixed the error I had made in the change-log entry so now also circle-ci passes.

**Comment by seberg** [comment]
I am happy to get it in, someone should maybe confirm they like it as well.  Am travelling right now, so may just punt for another while before doing another pass and merging, sorry.

**Comment by mattip** [comment]
Did you compare performance of calls to matmul and vecmat/matvec? I would be very surprised if they were different.

**Comment by mhvk** [comment]
There should be no performance difference - it uses the same code. The advantage of `matvec` and `vecmat` is that one can pass in stacks of vectors and stacks of matrices instead of being stuck to a single vector.

ping also @charris who originally suggested we have these functions...

**Comment by mattip** [comment]
We discussed this at a recent triage meeting and would like to put it in. It needs a final review.

**Review by seberg — APPROVED** [review]
I looked through, and the only thing that stood out a bit is that vecmat and matvec may have subtly different behavior when it comes to accuracy for float32.

Overall, still happy to put this in.

**Comment by mhvk** [comment]
@seberg - I hadn't thought about implementation differences for float32 at all...

I'm fairly sure I initially had `vecmat` quite similar to `matvec` here, for the blas call just ensuring the conjugate, but with a separate no-blas implementation that ensured conjugation of the vector (since one couldn't use the `matmat` one). Then, realizing that `vecdot` had more or less what was needed already, it just seemed simpler to use it, reducing the amount of extra code...

I do think it is better if the two functions behave the same way. My tendency would be to use the better precision sums for both, which would mean repeating the complex `vecdot` implementation without conjugation for `matvec`. That said, it is easier to imaging vector multiplication where the precision matters than matrix-vector, since the matrices would generally become quite huge.  

I'll look a bit at whether one can do this fairly easily, but let me know if you have suggestions...

**Comment by mhvk** [comment]
@seberg - after spending most of the day on this (grr... still do not quite get the blas dimension stuff - too much switching between fortran and C...), I ended up with a bit of a frankencode: now the blas matrix-vector `gemv` will be called for float arguments, but not for complex. In principle, it works fine for complex for `matvec`, but it cannot handle the conjugation of the vector in `vecmat`; for symmetry, I thought it might be best to just fall through to my default implementation in that case -- which uses the various `*_dot`, so can still be blas-optimized in some cases.

What do you think?

**Comment by seberg** [comment]
Sorry, I didn't mean to push you into a rabbit hole :/, just realized that there likely is a difference due to the discussion about sum and float32 and was wondering if we should worry.
I think you are right, `gemm` could be used to achieve this probably, but `gemv` doesn't seem to have the transposes we would need. 

I suspect your frankencode is fine (and the original would probably also have been fine).  @mattip should we just go ahead with this?

---

Pushed a commit to fix the `__module__` thing that failed CI in main.  Points at one other thing: I would be totally fine with only adding these to functions to `numpy.linalg` (no strong opinion, though).

**Comment by mhvk** [comment]
Thanks for pushing the fixes - I did squash them together so it is one commit again.

Given that `vecdot` and `matmul` are in the main namespace, I think these might as well too...

**Comment by seberg** [comment]
Yeah, I think we can just put this in for 2.2.  @charris wanted to branch this weekend. I pinged the mailing list; but if anyone shouts there (or here), we can always do a backport (it'll be a few weeks before the actual release anyway).

**Comment by mhvk** [comment]
With a fresh look, I now can get `gemm` to work. Also noticed a slight omission in the `blasable` tests. So, wait for merging just a little!

**Comment by mhvk** [comment]
OK, now did the gemm for vecmat complex - should be ready!

**Pull request merged**
